File: app/src/main/java/kr/co/yogiyo/rookiephotoapp/settings/BackupRestoreDialogFragment.java
package kr.co.yogiyo.rookiephotoapp.settings;

import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AlertDialog;
import android.support.v7.preference.PreferenceDialogFragmentCompat;
import android.util.Log;
import android.util.Pair;
import android.view.View;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.google.firebase.auth.FirebaseUser;

import org.reactivestreams.Publisher;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

import id.zelory.compressor.Compressor;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;
import kr.co.yogiyo.rookiephotoapp.Constants;
import kr.co.yogiyo.rookiephotoapp.GlobalApplication;
import kr.co.yogiyo.rookiephotoapp.R;
import kr.co.yogiyo.rookiephotoapp.diary.db.Diary;
import kr.co.yogiyo.rookiephotoapp.diary.db.LocalDiaryViewModel;
import kr.co.yogiyo.rookiephotoapp.settings.sync.DiaryBackupRestore;
import kr.co.yogiyo.rookiephotoapp.settings.sync.RestoredDiary;
import okhttp3.ResponseBody;

public class BackupRestoreDialogFragment extends PreferenceDialogFragmentCompat implements DialogInterface.OnShowListener {

    private static final String TAG = BackupRestoreDialogFragment.class.getSimpleName();
    private final String NO_DATA = "no_data";
    private final String NO_BACKUP_HISTORY = "Expected BEGIN_ARRAY but was BEGIN_OBJECT";
    private final String FAILED_TO_CONNECT = "Failed to connect";
    private final String ERROR_MESSAGE = "error_message";

    private SimpleDateFormat serverDateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault());

    private Compressor imageCompressor;

    private CompositeDisposable compositeDisposable = new CompositeDisposable();

    private DiaryBackupRestore diaryBackupRestore = new DiaryBackupRestore();

    private LocalDiaryViewModel localDiaryViewModel;

    private TextView askBackupRestoreText;
    private TextView cancelText;
    private TextView okText;

    private Context context;

    public static BackupRestoreDialogFragment newInstance() {
        return new BackupRestoreDialogFragment();
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        this.context = context;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        localDiaryViewModel = ViewModelProviders.of(this).get(LocalDiaryViewModel.class);
        imageCompressor = new Compressor(context);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        compositeDisposable.clear();
    }

    @Override
    protected void onBindDialogView(View view) {
        super.onBindDialogView(view);
        RelativeLayout backupRestoreDialogRelative = view.findViewById(R.id.relative_backup_restore_dialog);
        askBackupRestoreText = view.findViewById(R.id.text_ask_backup_restore);
        cancelText = view.findViewById(R.id.text_cancel);
        okText = view.findViewById(R.id.text_ok);
        ((SettingsActivity) context).addProgressBarInto(backupRestoreDialogRelative);

        String preferenceKey = getArguments().getString(SettingsActivity.PREFERENCE_KEY);
        switch (preferenceKey) {
            case SettingsActivity.BACKUP_DIALOG_KEY:
                askBackupRestoreText.setText(getString(R.string.text_ask_backup));
                okText.setOnClickListener(v -> {
                    setLoadingDialog();
                    executeBackup(GlobalApplication.getGlobalApplicationContext().getFirebaseAuth().getCurrentUser());
                });
                break;
            case SettingsActivity.RESTORE_DIALOG_KEY:
                askBackupRestoreText.setText(getString(R.string.text_ask_restore));
                okText.setOnClickListener(v -> {
                    setLoadingDialog();
                    executeRestore(GlobalApplication.getGlobalApplicationContext().getFirebaseAuth().getCurrentUser());
                });
                break;
        }
        cancelText.setOnClickListener(v -> dismiss());
    }

    @Override
    protected void onPrepareDialogBuilder(AlertDialog.Builder builder) {
        super.onPrepareDialogBuilder(builder);
        String preferenceKey = getArguments().getString(SettingsActivity.PREFERENCE_KEY);
        switch (preferenceKey) {
            case SettingsActivity.BACKUP_DIALOG_KEY:
                builder.setTitle(getString(R.string.text_do_backup));
                break;
            case SettingsActivity.RESTORE_DIALOG_KEY:
                builder.setTitle(getString(R.string.text_do_restore));
                break;
        }
        builder.setPositiveButton(null, null)
                .setNegativeButton(null, null);
    }

    @Override
    public void onDialogClosed(boolean positiveResult) {
        // Do nothing
    }

    // TODO : 백업/복원 실행할 때 다른 작업 못하도록 또는 바로 클릭 못하도록 또는 화면 분할 (작업 진행도 표시 필요할 듯)
    // TODO : 리사이징한 사진 복구 어떻게 할지 고민
    private void executeBackup(FirebaseUser currentUser) {
        compositeDisposable.add(
                localDiaryViewModel.findDiaries().toFlowable()
                        .subscribeOn(Schedulers.io())
                        .observeOn(Schedulers.io())
                        .map(diaries -> {
                            if (diaries.isEmpty()) {
                                throw new Exception(NO_DATA);
                            }

                            compositeDisposable.add(
                                    diaryBackupRestore.executePostClearDiary(currentUser)
                                            .subscribe(responseBody -> {
                                                // Do nothing
                                            }, throwable -> {
                                                // Do nothing
                                            })
                            );

                            return diaries;
                        })
                        .flatMapIterable(diaries -> diaries)
                        .flatMap((Function<Diary, Publisher<ResponseBody>>) diary -> {
                            if (diary.getImage() != null) {
                                File imageFile = new File(Constants.YOGIDIARY_PATH, diary.getImage());
                                if (imageFile.isFile()) {
                                    imageCompressor.setDestinationDirectoryPath(
                                            Constants.YOGIDIARY_PATH.getAbsolutePath() + File.separator + DiaryBackupRestore.COMPRESSED_FOLDER_NAME)
                                            .compressToFile(imageFile);
                                }
                            }

                            return diaryBackupRestore.executePostDiary(currentUser, diary);
                        })
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(responseDiariesBody -> {
                            Log.d(TAG, "body string " + responseDiariesBody.string());
                            if (responseDiariesBody.string().contains(ERROR_MESSAGE)) {
                                throw new Exception();
                            }
                        }, throwable -> {
                            ((SettingsActivity) context).hideLoading();
                            BackupRestoreDialogFragment.this.getDialog().dismiss();

                            if (NO_DATA.equals(throwable.getMessage())) {
                                ((SettingsActivity) context).createAlertDialog(context, null, getString(R.string.text_no_backup_data),
                                        getString(R.string.text_confirm), null, null, this).show();
                            } else if (throwable.getMessage().contains(FAILED_TO_CONNECT)) {
                                ((SettingsActivity) context).createAlertDialog(context, null, getString(R.string.text_network_error),
                                        getString(R.string.text_confirm), null, null, this).show();
                            } else {
                                ((SettingsActivity) context).createAlertDialog(context, null, getString(R.string.text_backup_fail),
                                        getString(R.string.text_confirm), null, null, this).show();
                            }

                            Log.d(TAG, throwable.getMessage());
                        }, () -> {
                            ((SettingsActivity) context).hideLoading();
                            BackupRestoreDialogFragment.this.getDialog().dismiss();

                            ((SettingsActivity) context).createAlertDialog(context, null, getString(R.string.text_backup_success),
                                    getString(R.string.text_confirm), null, null, this).show();
                        })
        );
    }

    private void executeRestore(FirebaseUser currentUser) {
        compositeDisposable.add(
                diaryBackupRestore.executeGetDiaries(currentUser)
                        .subscribeOn(Schedulers.io())
                        .observeOn(Schedulers.io())
                        .flatMapIterable(restoredDiaries -> {
                            if (restoredDiaries.isEmpty()) {
                                throw new Exception(NO_DATA);
                            }

                            return restoredDiaries;
                        })
                        .map(restoredDiary -> {
                            Date date = serverDateFormat.parse(restoredDiary.getDatetime());

                            String image = null;
                            if (restoredDiary.getImage() != null) {
                                image = restoredDiary.getImage().substring(restoredDiary.getImage().lastIndexOf(File.separator) + 1);
                            }

                            localDiaryViewModel.insertDiary(restoredDiary.getDiaryId(), date, image, restoredDiary.getDescription())
                                    .subscribe();

                            return restoredDiary;
                        })
                        .filter(restoredDiary -> restoredDiary.getImage() != null)
                        .flatMap((Function<RestoredDiary, Publisher<Pair<String, ResponseBody>>>) restoredDiary ->
                                diaryBackupRestore.executeGetImage(restoredDiary.getImage().substring(restoredDiary.getImage().indexOf(File.separator) + 1)))
                        .map(pair -> {
                            String imageFileName = pair.first.substring(pair.first.lastIndexOf(File.separator) + 1);
                            boolean writtenToDisk = BackupRestoreDialogFragment.this.writeResponseBodyToDisk(imageFileName, pair.second);

                            return writtenToDisk;
                        })
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(writtenToDisk -> Log.d(TAG, "success = " + writtenToDisk),
                                throwable -> {
                                    ((SettingsActivity) context).hideLoading();
                                    BackupRestoreDialogFragment.this.getDialog().dismiss();

                                    if (throwable.getMessage().equals(NO_DATA) || throwable.getMessage().contains(NO_BACKUP_HISTORY)) {
                                        ((SettingsActivity) context).createAlertDialog(context, null, getString(R.string.text_no_restore_data),
                                                getString(R.string.text_confirm), null, null, this).show();
                                    } else if (throwable.getMessage().contains(FAILED_TO_CONNECT)) {
                                        ((SettingsActivity) context).createAlertDialog(context, null, getString(R.string.text_network_error),
                                                getString(R.string.text_confirm), null, null, this).show();
                                    } else {
                                        ((SettingsActivity) context).createAlertDialog(context, null, getString(R.string.text_restore_fail),
                                                getString(R.string.text_confirm), null, null, this).show();
                                    }

                                    Log.d(TAG, throwable.getMessage());
                                }, () -> {
                                    ((SettingsActivity) context).hideLoading();
                                    BackupRestoreDialogFragment.this.getDialog().dismiss();

                                    ((SettingsActivity) context).createAlertDialog(context, null, getString(R.string.text_restore_success),
                                            getString(R.string.text_confirm), null, null, this).show();
                                })
        );
    }

    private boolean writeResponseBodyToDisk(String imageFileName, ResponseBody body) {
        if (!Constants.YOGIDIARY_PATH.exists()) {
            if (!Constants.YOGIDIARY_PATH.mkdirs()) {
                return false;
            }
        }

        File restoredImageFile = new File(Constants.YOGIDIARY_PATH, imageFileName);

        try (InputStream inputStream = body.byteStream();
             OutputStream outputStream = new FileOutputStream(restoredImageFile)) {
            byte[] fileReader = new byte[4096];

            long fileSize = body.contentLength();
            long fileSizeDownloaded = 0;

            int read = inputStream.read(fileReader);
            while (read != -1) {
                outputStream.write(fileReader, 0, read);
                fileSizeDownloaded += read;
                Log.d(TAG, "file download: " + fileSizeDownloaded + " of " + fileSize);

                read = inputStream.read(fileReader);
            }

            outputStream.flush();

            context.sendBroadcast((new Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE, Uri.fromFile(restoredImageFile))));

            return true;
        } catch (IOException e) {
            return false;
        }
    }

    private void setLoadingDialog() {
        ((SettingsActivity) context).showLoading();
        askBackupRestoreText.setText(getString(R.string.text_please_wait));
        cancelText.setVisibility(View.GONE);
        okText.setVisibility(View.GONE);
    }

    @Override
    public void onShow(DialogInterface dialog) {
        ((AlertDialog) dialog).getButton(AlertDialog.BUTTON_POSITIVE).setTextColor(ContextCompat.getColor(context, R.color.color_FF0000));
    }
}

File: app/src/main/java/kr/co/yogiyo/rookiephotoapp/settings/BackupRestoreDialogFragment.kt
package kr.co.yogiyo.rookiephotoapp.settings

import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.preference.PreferenceDialogFragmentCompat
import android.util.Log

import com.google.firebase.auth.FirebaseUser
import com.google.gson.JsonParser

import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.Locale

import id.zelory.compressor.Compressor
import io.reactivex.Flowable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.functions.BiFunction
import io.reactivex.schedulers.Schedulers
import kr.co.yogiyo.rookiephotoapp.Constants
import kr.co.yogiyo.rookiephotoapp.GlobalApplication
import kr.co.yogiyo.rookiephotoapp.R
import kr.co.yogiyo.rookiephotoapp.diary.db.Diary
import kr.co.yogiyo.rookiephotoapp.diary.db.LocalDiaryViewModel
import kr.co.yogiyo.rookiephotoapp.settings.sync.DiaryBackupRestore
import okhttp3.ResponseBody

class BackupRestoreDialogFragment : PreferenceDialogFragmentCompat() {

    private val serverDateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault())

    private val imageCompressor by lazy { Compressor(context) }

    private val compositeDisposable = CompositeDisposable()

    private val diaryBackupRestore = DiaryBackupRestore()

    private val localDiaryViewModel by lazy {
        ViewModelProviders.of(this).get(LocalDiaryViewModel::class.java)
    }

    override fun onDestroy() {
        super.onDestroy()
        compositeDisposable.clear()
    }

    override fun onPrepareDialogBuilder(builder: AlertDialog.Builder?) {
        super.onPrepareDialogBuilder(builder)
        arguments?.let { arguments ->
            when (arguments.getString(SettingsActivity.PREFERENCE_KEY)) {
                SettingsActivity.BACKUP_DIALOG_KEY -> {
                    builder?.run {
                        setTitle(R.string.text_do_backup)
                        setMessage(R.string.text_ask_backup)
                    }
                }
                SettingsActivity.RESTORE_DIALOG_KEY -> {
                    builder?.run {
                        setTitle(R.string.text_do_restore)
                        setMessage(R.string.text_ask_restore)
                    }
                }
            }
            builder?.run {
                setPositiveButton(R.string.text_dialog_ok, null)
                setNegativeButton(R.string.text_dialog_no) { _, _ -> dismiss() }
            }
        }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        dialog.setOnShowListener { dialog ->
            arguments?.let { arguments ->
                (dialog as AlertDialog).getButton(AlertDialog.BUTTON_POSITIVE).setOnClickListener {
                    setLoadingDialog(true)
                    when (arguments.getString(SettingsActivity.PREFERENCE_KEY)) {
                        SettingsActivity.BACKUP_DIALOG_KEY -> {
                            executeBackup(GlobalApplication.globalApplicationContext.firebaseAuth.currentUser)
                        }
                        SettingsActivity.RESTORE_DIALOG_KEY -> {
                            executeRestore(GlobalApplication.globalApplicationContext.firebaseAuth.currentUser)
                        }
                    }
                }
            }
        }
    }

    override fun onDialogClosed(positiveResult: Boolean) {
        // Do nothing
    }

    // TODO : 백업/복원 실행할 때 다른 작업 못하도록 또는 바로 클릭 못하도록 또는 화면 분할 (작업 진행도 표시 필요할 듯)
    // TODO : 리사이징한 사진 복구 어떻게 할지 고민
    private fun executeBackup(currentUser: FirebaseUser?) {
        compositeDisposable.add(
                localDiaryViewModel.findDiaries().toFlowable()
                        .subscribeOn(Schedulers.io())
                        .observeOn(Schedulers.io())
                        .flatMap { diaries ->
                            if (diaries.isEmpty()) {
                                throw Exception(NO_DATA)
                            } else {
                                val flowableDiaries = Flowable.just(diaries)
                                Flowable.zip(
                                        flowableDiaries,
                                        diaryBackupRestore.executePostClearDiary(currentUser!!),
                                        BiFunction<List<Diary>,
                                                ResponseBody,
                                                Pair<Flowable<List<Diary>>, ResponseBody>> { _, responseBody ->
                                            Pair(flowableDiaries, responseBody)
                                        }
                                )
                            }
                        }
                        .flatMap {
                            if (JsonParser().parse(it.second.string()).asJsonObject
                                            .get(SERVER_RESPONSE_KEY).asString != SERVER_RESPONSE_CLEAR_SUCCESS) {
                                throw Exception(FAILED_TO_CONNECT)
                            } else {
                                it.first
                            }
                        }
                        .flatMapIterable { diaries -> diaries }
                        .flatMap { diary ->
                            if (diary.image != null) {
                                val imageFile = File(Constants.FOONCARE_PATH, diary.image)
                                if (imageFile.isFile) {
                                    imageCompressor.setDestinationDirectoryPath(
                                            Constants.FOONCARE_PATH.absolutePath + File.separator + DiaryBackupRestore.COMPRESSED_FOLDER_NAME)
                                            .compressToFile(imageFile)
                                }
                            }

                            diaryBackupRestore.executePostDiary(currentUser!!, diary)
                        }
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe({ responseDiariesBody ->
                            Log.d(TAG, "body string " + responseDiariesBody.string())
                            if (responseDiariesBody.string().contains(ERROR_MESSAGE)) {
                                throw Exception()
                            }
                        }, { throwable ->
                            val errorMessageResource = throwable.message?.let { message ->
                                when {
                                    message == NO_DATA -> R.string.text_no_backup_data
                                    message.contains(FAILED_TO_CONNECT) -> R.string.text_network_error
                                    else -> R.string.text_backup_fail
                                }
                            } ?: R.string.text_backup_fail

                            (context as SettingsActivity).createAlertDialog(context!!, null, getString(errorMessageResource),
                                    getString(R.string.text_confirm), null, null, null).show()

                            setLoadingDialog(false)

                            Log.d(TAG, throwable.message)
                        }, {
                            (context as SettingsActivity).createAlertDialog(context!!, null, getString(R.string.text_backup_success),
                                    getString(R.string.text_confirm), null, null, null).show()

                            setLoadingDialog(false)
                        })
        )
    }

    private fun executeRestore(currentUser: FirebaseUser?) {
        compositeDisposable.add(
                diaryBackupRestore.executeGetDiaries(currentUser!!)
                        .subscribeOn(Schedulers.io())
                        .observeOn(Schedulers.io())
                        .flatMapIterable { restoredDiaries ->
                            if (restoredDiaries.isEmpty()) {
                                throw Exception(NO_DATA)
                            }

                            restoredDiaries
                        }
                        .map { restoredDiary ->
                            val date = serverDateFormat.parse(restoredDiary.datetime)
                            val image: String? = restoredDiary.image?.substring(restoredDiary.image!!.lastIndexOf(File.separator) + 1)

                            localDiaryViewModel.insertDiary(restoredDiary.diaryId, date, image, restoredDiary.description)
                                    .subscribe()

                            restoredDiary
                        }
                        .filter {
                            it.image != null
                        }
                        .flatMap { restoredDiary ->
                            diaryBackupRestore.executeGetImage(restoredDiary.image!!.substring(restoredDiary.image!!.indexOf(File.separator) + 1))
                        }
                        .map { pair ->
                            val imageFileName = pair.first.substring(pair.first.lastIndexOf(File.separator) + 1)
                            val writtenToDisk = this@BackupRestoreDialogFragment.writeResponseBodyToDisk(imageFileName, pair.second)

                            writtenToDisk
                        }
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe({ writtenToDisk -> Log.d(TAG, "success = $writtenToDisk") },
                                { throwable ->
                                    val errorMessageResource = throwable.message?.let { message ->
                                        when {
                                            message == NO_DATA || message.contains(NO_BACKUP_HISTORY) -> R.string.text_no_restore_data
                                            message.contains(FAILED_TO_CONNECT) -> R.string.text_network_error
                                            else -> R.string.text_restore_fail
                                        }
                                    } ?: R.string.text_restore_fail

                                    (context as SettingsActivity).createAlertDialog(context!!, null, getString(errorMessageResource),
                                            getString(R.string.text_confirm), null, null, null).show()

                                    setLoadingDialog(false)

                                    Log.d(TAG, throwable.message)
                                }, {
                            (context as SettingsActivity).createAlertDialog(context!!, null, getString(R.string.text_restore_success),
                                    getString(R.string.text_confirm), null, null, null).show()

                            setLoadingDialog(false)
                        })
        )
    }

    private fun writeResponseBodyToDisk(imageFileName: String, body: ResponseBody): Boolean {
        if (!Constants.FOONCARE_PATH.exists()) {
            if (!Constants.FOONCARE_PATH.mkdirs()) {
                return false
            }
        }

        val restoredImageFile = File(Constants.FOONCARE_PATH, imageFileName)

        try {
            body.byteStream().use { inputStream ->
                FileOutputStream(restoredImageFile).use { outputStream ->
                    val fileReader = ByteArray(4096)

                    val fileSize = body.contentLength()
                    var fileSizeDownloaded: Long = 0

                    var read = inputStream.read(fileReader)
                    while (read != -1) {
                        outputStream.write(fileReader, 0, read)
                        fileSizeDownloaded += read.toLong()
                        Log.d(TAG, "file download: $fileSizeDownloaded of $fileSize")

                        read = inputStream.read(fileReader)
                    }

                    outputStream.flush()

                    context!!.sendBroadcast(Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE, Uri.fromFile(restoredImageFile)))

                    return true
                }
            }
        } catch (e: IOException) {
            return false
        }
    }

    private fun setLoadingDialog(visible: Boolean) {
        (context as SettingsActivity).showLoading(visible)
        if (visible) {
            dialog.hide()
        } else {
            dismiss()
        }
    }

    companion object {

        private val TAG = BackupRestoreDialogFragment::class.java.simpleName
        private const val NO_DATA = "no_data"
        private const val NO_BACKUP_HISTORY = "Expected BEGIN_ARRAY but was BEGIN_OBJECT"
        private const val FAILED_TO_CONNECT = "Failed to connect"
        private const val ERROR_MESSAGE = "error_message"
        private const val SERVER_RESPONSE_KEY = "message"
        private const val SERVER_RESPONSE_CLEAR_SUCCESS = "clear success"

        fun newInstance(): BackupRestoreDialogFragment {
            return BackupRestoreDialogFragment()
        }
    }
}